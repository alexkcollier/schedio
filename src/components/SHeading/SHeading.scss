.heading {
  font-weight: $font-weight-bold;
  margin-top: 2rem;
  margin-bottom: 1rem;
  text-transform: uppercase;

  &--level-1 {
    @include font-size(3);
  }

  &--level-2 {
    @include font-size(4);
  }

  &--level-3 {
    @include font-size(5);
  }

  &--level-4 {
    @include font-size(5);
    text-transform: none;
  }

  &--display {
    &-1 {
      @include font-size(1);
    }

    &-2 {
      @include font-size(2);
    }
  }

  &--subtle {
    font-weight: $font-weight-light;
    text-transform: none;
  }

  &--no-case {
    text-transform: none;
  }

  &--title {
    margin-bottom: 0;
  }

  &--subtitle {
    margin-top: 0 !important;
  }

  & + & {
    margin-top: 1rem;
  }

  &:first-child {
    margin-top: 0;
  }
}
