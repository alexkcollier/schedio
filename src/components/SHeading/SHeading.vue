<template>
  <component
    :is="componentTag"
    :class="classList"
  >
    <slot />
  </component>
</template>

<script>
export default {
  name: 'SHeading',

  props: {
    level: {
      type: [String, Number],
      default: 1,
      validator: value => {
        return (
          Number(value) <= 4 ||
          console.error('Level must be 4 or lower. 2 or lower if `isDisplay`.')
        )
      }
    },

    isDisplay: {
      type: Boolean,
      default: false
    },

    hasNoCase: {
      type: Boolean,
      default: false
    },

    isSubtle: {
      type: Boolean,
      default: false
    },

    isTitle: {
      type: Boolean,
      default: false
    },

    isSubtitle: {
      type: Boolean,
      default: false
    },

    tag: {
      type: String,
      default: '',
      validator: value => {
        const allowedTags = ['div', 'span', 'a', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6']

        return (
          // allow empty
          !value ||
          allowedTags.includes(value) ||
          console.error(`Tag must be one of: ${allowedTags.join(', ')}.`)
        )
      }
    }
  },

  computed: {
    classList() {
      return [
        'heading',
        {
          [`heading--level-${this.level}`]: this.level,
          'heading--display': this.isDisplay,
          'heading--subtle': this.isSubtle,
          'heading--no-case': this.hasNoCase,
          'heading--title': this.isTitle,
          'heading--subtitle': this.isSubtitle
        }
      ]
    },

    componentTag() {
      return this.tag || 'h' + this.level
    }
  }
}
</script>
