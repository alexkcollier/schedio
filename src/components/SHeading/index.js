import SHeading from './SHeading.vue'

export { SHeading }

export default { install: Vue => Vue.component(SHeading.name, SHeading) }
