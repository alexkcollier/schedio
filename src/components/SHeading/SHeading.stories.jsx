// Global registration doesn't work with `jsx` templates
import { SHeading } from '@/components/SHeading'
import PropList from '@/docs/PropList'
import StoryContainer from '@/docs/StoryContainer'
import { boolean, radios, select, text, withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'

const headingLevels = ['1', '2', '3', '4']
const displayLevels = ['1', '2']
const options = radios(
  'Pick one',
  ['Default', 'has-no-case', 'is-subtle'],
  'Default',
  'Optional props'
)
const tagOptions = ['', 'div', 'span', 'a']

storiesOf('Components/SHeading', module)
  .addDecorator(withKnobs)
  .add('Heading', function Heading() {
    return {
      props: {
        props: {
          default: {
            level: select('level', headingLevels, '1', 'Optional props'),
            displayLevel: select('display-level', displayLevels, '1', 'Optional props'),
            hasNoCase: options === 'has-no-case',
            isSubtle: options === 'is-subtle',
            isDisplay: boolean('is-display', false, 'Optional props'),
            isTitle: boolean('is-title', false, 'Optional props'),
            isSubtitle: boolean('is-subtitle', false, 'Optional props'),
            tag: select('tag', tagOptions, '', 'Optional props')
          }
        },
        slotContent: {
          default: text('Default', 'Slot content', 'Slots')
        }
      },
      render(h) {
        const { props } = this.$props

        return (
          <StoryContainer>
            <SHeading>Heading</SHeading>
            <p>
              <code>SHeading</code> defaults to using <code>h1-h6</code> tags, but can be overriden
              to be a <code>{'<div>'}</code> or <code>{'<span>'}</code> instead using the{' '}
              <code>tag</code> prop.
            </p>

            <SHeading level="2">Example</SHeading>
            <div>
              <SHeading {...{ props }}>
                Level {props.level}: {this.slotContent}
              </SHeading>
            </div>

            <PropList component={SHeading} />
          </StoryContainer>
        )
      }
    }
  })
  .add('Usage: Headings', function UsageHeadings() {
    return {
      props: {
        props: {
          default: {
            level: select('level', headingLevels, '1', 'Optional props'),
            hasNoCase: options === 'has-no-case',
            isSubtle: options === 'is-subtle',
            isTitle: boolean('is-title', false, 'Optional props'),
            isSubtitle: boolean('is-subtitle', false, 'Optional props')
          }
        },
        slotContent: {
          default: text('Default', 'Slot content', 'Slots')
        }
      },
      render(h) {
        const { slotContent, props } = this.$props

        return (
          <StoryContainer>
            <SHeading>Usage: Heading</SHeading>
            <p>Headings help identify key sections and information within a document.</p>

            <SHeading level="2">Example</SHeading>
            <div>
              <SHeading {...{ props }}>
                Level {props.level}: {slotContent}
              </SHeading>
            </div>

            <SHeading level="3">Consecutive Headings</SHeading>
            <p>Headings placed next to each other will be spaced 1rem apart.</p>
            <div>
              <SHeading>Title</SHeading>
              <SHeading level="2">Subtitle</SHeading>
            </div>

            <div>
              <SHeading isTitle>Title</SHeading>
              <SHeading isSubtitle level="2">
                Subtitle
              </SHeading>
            </div>

            <PropList component={SHeading} />
          </StoryContainer>
        )
      }
    }
  })
  .add('Usage: Display', function UsageDisplay() {
    const options = radios(
      'Pick one',
      ['Default', 'has-no-case', 'is-subtle'],
      'Default',
      'Optional'
    )

    return {
      props: {
        props: {
          default: {
            level: select('level', headingLevels, '1', 'Optional props'),
            displayLevel: select('display-level', displayLevels, '1', 'Optional props'),
            hasNoCase: options === 'has-no-case',
            isSubtle: options === 'is-subtle',
            isDisplay: true,
            isTitle: boolean('is-title', false, 'Optional props'),
            isSubtitle: boolean('is-subtitle', false, 'Optional props'),
            tag: select('tag', tagOptions, '', 'Optional props')
          }
        },
        slotContent: {
          default: text('Default', 'Slot content', 'Slots')
        }
      },
      render(h) {
        const { slotContent, props } = this.$props

        return (
          <StoryContainer>
            <SHeading>Usage: Display</SHeading>
            <p>
              Headings are key to structure, and displays are decorative. Displays should be used
              sparingly, typically once per document.
            </p>

            <SHeading level="2">Example</SHeading>
            <div>
              <SHeading {...{ props }}>
                Level {props.level}: {slotContent}
              </SHeading>
            </div>

            <PropList component={SHeading} />
          </StoryContainer>
        )
      }
    }
  })
