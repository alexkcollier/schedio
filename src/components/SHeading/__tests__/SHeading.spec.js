import { shallowMount } from '@vue/test-utils'
import SHeading from '@/components/SHeading/SHeading.vue'

describe('SHeading.vue', () => {
  let errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => {})
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(SHeading, {
      propsData: {
        level: 1
      },
      slots: {
        default: 'Heading Content'
      }
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can have the tag specified', () => {
    wrapper.setProps({ tag: 'div' })
    expect(wrapper.props('tag')).toBe('div')
  })

  it('should compute the h tag to use', () => {
    for (let i; i <= 4; i++) {
      shallowMount(SHeading, { propsData: { level: i } })
      expect(wrapper.props('tag')).toBe('h' + i)
    }
  })

  it('should have a level of 4 or less', () => {
    for (let i; i <= 4; i++) {
      shallowMount(SHeading, { propsData: { level: i } })
      expect(errorSpy).not.toBeCalled()
    }

    shallowMount(SHeading, { propsData: { level: 5 } })
    expect(errorSpy).toBeCalled()
  })

  it('can be display', () => {
    wrapper.setProps({ isDisplay: true })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can have no case', () => {
    wrapper.setProps({ hasNoCase: true })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can be subtle', () => {
    wrapper.setProps({ isSubtle: true })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
