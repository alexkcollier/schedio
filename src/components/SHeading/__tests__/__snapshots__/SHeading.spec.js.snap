// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SHeading.vue can be display 1`] = `<h1 class="heading heading--level-1 heading--display">Heading Content</h1>`;

exports[`SHeading.vue can be subtle 1`] = `<h1 class="heading heading--level-1 heading--subtle">Heading Content</h1>`;

exports[`SHeading.vue can have no case 1`] = `<h1 class="heading heading--level-1 heading--no-case">Heading Content</h1>`;

exports[`SHeading.vue renders correctly 1`] = `<h1 class="heading heading--level-1">Heading Content</h1>`;
